### DISCUSSION TITLE
reimplement the `twophase` controller in chaos mesh

### DISCUSSION THREAD

**Discussion by YangKeao** [chaos-mesh/chaos-mesh#1520] [General]
It has been one year since Chaos Mesh got open sourced. The `controller-manager` of Chaos Mesh keeps growing and turning into a huge and hard to manage project :cry: . The `twophase` controller of it epitomises the difficulty, which I have submitted a lot of PRs to try to make it simpler.

In this discussion, I will do a self-criticism about the bad design —— `duration` and `scheduler` of current controller and CRDs and will also explain the historical reason of the bad design. Several choices to overcome these issues will be provided as references (with or without breaking the compatibility). 

Under this discussion, **I hope we can decide one acceptable plan to make this thing better. And the implementation of the choice should arrive as soon as possible. **

The `duration` and `scheduler` fields in every CRDs is really a bad design. It combines the thing of `scheduler` with "chaos implementation". With these fields in the CRD, we have to manage much more status in one CRD. What's more, a lot of "high-level" features are brought to them later: like `pause` and `dynamic scheduler`.

Now, nearly every new features on the controller (e.g. "injecting chaos to newly created chaos", "dynamically change configuration") are blocked by the complexity of `twophase` controller. We (at least I) didn't have enough confidence to write a bug-free implementation without breaking `twophase` controller.

Though I have tried to make `twophase` controller easier, it has been proved this attempt failed finally :crying_cat_face:. It doesn't make things better enough.

### historical reason

The first CRD of Chaos Mesh is `PodChaos`, and the first chaos of Chaos Mesh is `pod-kill`, which is a "job" but not a "status",. Only with `duration` and `scheduler`, which means the pods are killed periodically, it becomes a status. Then all of the Chaos followed it, without thinking whether it's really a good design.

### Solution

All these solutions will make the CRD more complicate. Just like `Deployment -> ReplicaSet -> Pod`, spliting the functions into multiple standalone resource shouldn't be feared, and cloud be a good practice. The only concern is about how to describe the phenomenon to the users.

1. Wait for `CronWorkflow` and remove these two fields from the CRD
Pros: 
    * The design could be much simpler, and no extra work need to do.
Cons:
    * Break compatibility
    * It will take several months before it arrives :cry: (Though I'm willing to help and make it arrive earlier :yum: )
2. Treat CRD with `duration` and `scheduler` like a standalone resource, which will create common chaos to fulfill its implementation. #1481 
Pros:
    * The code base is relatively small, and easy to understand
    * It could be a practice to manager multi-level custom resources, which us developers should be able to do.
Cons:
    * Bring confusing user experience: the users will find a lot of `XXXChaos` created by the machine.
    * Didn't eliminate the multi-meaning of `XXXChaos`
3. Create a new unstructured CRD `RawChaos`, and the twophase controller implements through creating and monitoring the `RawChaos`.
Pros:
    * A one-layer `RawChaos` CRD will bring us a unified interface of chaos, which could be really helpful for `plugin`
    * A newly created CRD will not break the compatibility :yum: 
Cons:
    * The function of `RawChaos` and common chaos are nearly the same.

@Yiyiyimu @YangKeao @STRRL @Hexilee You have all contributed to the twophase controller or talked about the design of it with me before. Do you have any suggestions on this problem?

**Comment by Colstuwjx** [comment]
I have participated in the discussion of "update scheduler", see [discussion 1491 #comment](https://github.com/chaos-mesh/chaos-mesh/discussions/1491#discussioncomment-340421), and I'd like to give some suggestions from my view:

1. What's the difference between CronWorkflow and Workflow with Cron scheduler? If they are the same thing, I suggest that we make a separate component called Chaos Scheduler, and it cares about the one-time job or cron-like job scheduling, similar to your plan 2 ;

2. In general, our chaos didn't equal a job, if we said ran a job, the job should have a status that indicates it succeeds or failed. **But Chaos didn't**. We preset each chaos would keep running if it's not paused, and it doesn't have an exit gate. **We may need to think that how to define a Chaos Job, its primitive and when to exit**, maybe it could combine with your plan 3 **RawChaos**, or sth called **MetaChaos**;

3. For the dev roadmap, I suggest that we prepare to land this feature in version 2.0, and deprecate the old version at a scheduled date ( Prometheus project refactored its storage engine from 1.0 to 2.0, and it finally merged into the mainstream in [this PR](https://github.com/prometheus/prometheus/pull/5805), I think it should be a good example for a breaking change ).

Thanks.

**Reply by YangKeao** [reply]
For the first question: 

@STRRL I don't think there will be a `Workflow` with `Cron` scheduler. Every `Workflow` will only be executed once. The `CronWorkflow` is used to create `Workflow` periodically. The `argo` also split these two features into `Workflow` and `CronWorkflow`. I think it's a good practice (as combining `scheduler` with other CRD has been proved to be a bad idea :imp: )

For the second question:

:+1: Yes. This is an essential problem :cry:. Some chaos are more like a "status", which infers that the cluster is / (should be) wrong, e.g. the network connection is blocked, the disk is degraded... And I believe these "status" matches the philosophy of Kubernetes, just like `ReplicaSet` describes the status that there is / (should be) 5 pods in the cluster.

However, there is another type of Chaos, like `pod-kill`, which is not a "status" but more like a `job` :crying_cat_face:

**Reply by STRRL** [reply]
`CronWorkflow` is much simpler than `Workflow` with `Cron`; `CronWorkflow` only creates each new `Workflow` and leaves it runs alone.

Now we always use `cron` and `duration` as one bound bundle, so the controller should keep tracking the status of the workload, it's what the two-phase controller does.

**Comment by STRRL** [comment]
I want to break the compatibility and make our definition simple. I think we should not fear breaking backward compatibility. 

Create new `CronIoChaos`, `CronPodChaos`, etc. **Move** the fields: `duration` and `scheduler` to the new `CronXXXChaos`(remove them from current CRD), then we could apply the two-phase controller on it and reuse the common controller.

**Reply by YangKeao** [reply]
I prefer a `CronChaos`, `CronWorkflow` or `Schedule` (no matter how you name it) single CRD directly, rather than create `CronXXXChaos` for every `XXXChaos` (though they could be generated).

**Reply by YangKeao** [reply]
@cwen0 Can we tolerate this kind of compatibility-break? If we do, I prefer this solution too.

We can enable these CRDs in 1.x and remove `scheduler` and `duration` in 2.x

**Reply by cwen0** [reply]
> @cwen0 Can we tolerate this kind of compatibility-break? If we do, I prefer this solution too.
> 
> We can enable these CRDs in 1.x and remove `scheduler` and `duration` in 2.x

It is OK. We should keep compatibility in 1.x version and import this change to 2.x version.

**Reply by YangKeao** [reply]
In case that we could use new CRD to handle this problem, I have a new idea :smile_cat: .

Why don't we use a single `Schedule` CRD to do the both jobs for `CronXXXChaos` and `CronWorkflow`? They are doing the same thing: creating resources periodically and handle the status. The only difference is that one is to create `XXXChaos` and the other is to create `Workflow`. @STRRL What's your opinion?

(Maybe `Schedule` is not a good name. Do you have any suggestions on a better name?)

**Reply by STRRL** [reply]
The current two-phase controller handles both `cron` and `duration`. I  think `Schedule` could only resolve `cron`, and the `duration` is not covered by the common controller. 
Where do we put the codes about "delete that CRD after `duration`"?

**Reply by Colstuwjx** [reply]
@YangKeao Yes, can't agree more with the idea of `use a single CRD for scheduling`.

@STRRL I think we can split the two-phase into two part of features: 1. cron schedule, 2. duration and pause(actually it should be **exit**) , I'd like to share more details in my mind:

1. simplify the chaos. Let's said it's a `MetaChaos` for a chaos unit, and it should only support running with a certain interval and duration for a single run. Each business level chaos, like PodChaos or NetworkChaos, base on the `MetaChaos`. BTW, if we take this solution, We may need to refactor the current implementation of PodFailure\PodKill\ContainerKill etc, make them built on top of `MetaChaos`;

2. make each chaos that has a beginning also have an end. In another word, each chaos becomes a **job**. For example, a pod kill chaos with `duration: 30s, interval: 10s`, should run no more than 3 times, and after it met the duration time, it should become **finished** state. The **state** chaos, like NetworkChaos, is the same too. For example, a network loss chaos with `duration: 30s, interval: 10s`, should be `working10s-sleep10s-working10s`, and it also finally exit with **finished** state;

3. separate a component for scheduling, and it's a replacement for two-phase controller, e.g. call it `ChaosSchedule`. It should support run business chaos in `now`, `once` or `cron-like` mode ( xref: [Litmus Scheduling Page](https://docs.litmuschaos.io/docs/scheduling/) );

4. remove two-phase controller since we now have `MetaChaos` and `ChaosSchedule`;

5. think about the difference between CronWorkflow and Workflow with `ChaosSchedule`. IMO, I think we **ONLY** need Workflow for controlling the `if-then-else-then` logic since `ChaosSchedule` did the cron and ad-hoc schedule work.

**Reply by Colstuwjx** [reply]
> Where do we put the codes about "delete that CRD after duration"?

@STRRL if we take my suggested `MetaChaos with ChaosSchedule` solution, I think each chaos should become a **job** definition, and the relationship between `ChaosSchedule` and `Chaos` just like Kubernetes `Cronjob` with `Job`, we could make a field like `jobCleanUpPolicy` inside `ChaosSchedule` spec, and let it clean history jobs.

**Reply by STRRL** [reply]
I still feel confused about `MetaChaos` @Colstuwjx. Could you introduce more about its fields or behavior?

For example, if I submit a NetworkChaos that affects 20 seconds each 5 minutes, what happened with `Schdule`, `MetaChaos` and `NetworkChaos`?

**Reply by YangKeao** [reply]
> For example, a network loss chaos with `duration: 30s, interval: 10s`, should be `working10s-sleep10s-working10s`, and it also finally exit with finished state;

However, what if I want it to be `working10s-sleep5s-working10s`? It seems impossible for `duration` and `interval`? And the `working10s-sleep10s-working10s` is also confusing for me :crying_cat_face:

**Reply by Colstuwjx** [reply]
`MetaChaos` is just a code-level word, it's similar to [common](https://github.com/chaos-mesh/chaos-mesh/blob/master/controllers/common/types.go), the difference is that `MetaChaos` implements `duration` and `interval`, and each business level chaos could inherit from `MetaChaos` and fill out the actual reconcile code by themselves.

For example, if a user submit a NetworkChaos that affects 20 seconds each 5 minutes, its spec should be sth like below:

```
apiVersion: chaos-mesh.org/v1alpha1
kind: NetworkChaos
metadata:
  name: network-partition-example
  namespace: chaos-testing
spec:
  action: partition
  mode: one
  selector:
    labelSelectors:
      'app.kubernetes.io/component': 'tikv'
  direction: to
  target:
    selector:
      namespaces:
        - tidb-cluster-demo
      labelSelectors:
        'app.kubernetes.io/component': 'tikv'
    mode: one
  interval: '280s'
  duration: '300s'
```

*ps. this duration indicates the total running duration of the chaos, after it reached the duration time, chaos would update its status to **finished**, and it would never start again*

By default, the NetworkChaos would run at the create time, and it lasts 20s, so it should be `working20s-cold280s-exit`. The difference between MetaChaos and current implementation is that, MetaChaos controller handled the `interval` and `duration`, but it doesn't support complicate schedule like two-phase controller did. It also let each chaos has a chance to exit. For now, our chaos would never exit, **unless we delete it**.

Each chaos built on top of `MetaChaos` could work independently. It could also work with `ChaosScheduler`. For example, if we'd like to run chaos at a specified time, we could create a `ChaosSchedule` instance like below:

```
apiVersion: chaos-mesh.org/v1alpha1
kind: ChaosSchedule
metadata:
  name: schedule-network-chaos-at-specified-time
spec:
  schedule:
    once:
      executionTime: "2020-05-12T05:47:00Z"   
 chaosSpec:
    spec:
        kind: NetworkChaos
        action: partition
        mode: one
        selector:
            labelSelectors:
            'app.kubernetes.io/component': 'tikv'
        direction: to
        target:
            selector:
            namespaces:
                - tidb-cluster-demo
            labelSelectors:
                'app.kubernetes.io/component': 'tikv'
            mode: one
        interval: '280s'
        duration: '300s'
```

**Reply by Colstuwjx** [reply]
> However, what if I want it to be working10s-sleep5s-working10s? It seems impossible for duration and interval? And the working10s-sleep10s-working10s is also confusing for me

This duration is `total duration of the chaos`, a.k.a chaos lifetime, so we shouldn't assume the chaos would run the loop forever. In your case, it should be `total duration: 25s, interval: 5s`.

And yes, it's hard to cover all cases by the `MetaChaos`, e.g. cron schedule. So we need `MetaChaos` works with `ChaosSchedule`, rather than make these logical all into the two-phase controller.

**Reply by YangKeao** [reply]
This discussion is about to remove `cron` related things from current `XXXChaos` to simplify the controller. It seems that the `interval` and `duration` is as complicated as how it is now :crying_cat_face:

**Reply by YangKeao** [reply]
@Colstuwjx Sorry. I don't understand. Why `duration: 25s, interval: 5s` is `working10s-sleep5s-working10s`, but not `working5s-sleep5s-working5s-sleep5s-working5s`? Where would I specify the `working` time?

**Reply by Colstuwjx** [reply]
> @Colstuwjx Sorry. I don't understand. Why duration: 25s, interval: 5s is working10s-sleep5s-working10s, but not working5s-sleep5s-working5s-sleep5s-working5s? Where would I specify the working time?

Yes, my mistake, it seems that this design still has some bugs 😓

**Reply by STRRL** [reply]
> The difference between MetaChaos and current implementation is that, MetaChaos controller handled the interval and duration, but it doesn't support complicate schedule like two-phase controller did.

I think both of them are the same. 🙈 It also needs to manage the states of `Running` or `Waiting` and set a timer to change one to another.

> It also let each chaos has a chance to exit.  For now, our chaos would never exit, unless we delete it.

Yeah, I think that's a good feature. We suppose the user could make this scenario by workflow.


> The current two-phase controller handles both cron and duration. I think Schedule could only resolve cron, and the duration is not covered by the common controller.
Where do we put the codes about "delete that CRD after duration"?

Actually, I do not think there is somewhere that be suited to that logic. So let's just make another place to write these codes!🤪

There are two types of chaos, one is like `NetworkChaos`, which could affect for a duration. Another is like `AwsChaos`, which only executes one operation.

**TLDR**; What I do is build a new concept: `ChaosPeriod` which holds the `duration`. I want to use 2-layer(Schedule/Chaos) for `AwsChaos`, and 3-layer(Schedule/ChaosPeriod/Chaos) for `NetworkChaos`. Another thing is I want to keep the historical Chaos, so the way to recover is appending a new field that represents it's already finished instead of deleting the CRD object.



Just like 2-layer structure, `Schedule -> Chaos`:

```
podkill-every-one-minutes
- podkill-every-one-minutes-pfgke(with status Completed: Completed)
- podkill-every-one-minutes-ad92k(with status Completed: Completed)
- podkill-every-one-minutes-2ms8o(with status Completed: Uncompleted)
```

> Although pod kill should execute so-fast, `podkill-every-one-minutes-2ms8o` still in `Uncompleted` means it just created and the operation is still under execuing. 😜

3-layer-structure, `Schedule -> ChaosPeriod -> Chaos`

```
iochaos-5-mins-in-each-8-mins
- iochaos-5-mins-in-each-8-mins-ak3mx(Here is `ChaosPeriod`, It determine that when the chaos should be finished)
  - iochaos-5-mins-in-each-8-mins-ak3mx-2mxus(with spec: Phase: Finished)
- iochaos-5-mins-in-each-8-mins-ws2m8
  - iochaos-5-mins-in-each-8-mins-ws2m8-92ms8(with spec: Phase: Effecting)
```

> The spec of `iochaos-5-mins-in-each-8-mins-ws2m8-92ms8` holds the expected state of the chaos experiments. And the chaos controller keeps doing something to make it come true.

> Since the spec of `iochaos-5-mins-in-each-8-mins-ws2m8-92ms8` holds the expected state, so the `spec: Phase: Effecting/Finished` makes sense.

I think `Schedule` could work well with `AwsChaos` directly, but `Schedule` will create more than one `AwsChaos` and I want to keep the historical `AwsChaos`( like `CronJob` will keep `successfulJobsHistoryLimit`/`failedJobsHistoryLimit` historical `Jobs`). I think `AwsChaos` should add a status field called "Completed" that presents the status of the current operation,  its available values are `Uncompleted`, `Completed`, and `Failed`.

After that, we could use `Schedule` to schedule `AwsChaos`, `PodChaos` with a 2-layer structure. 

For example, I want to kill a Pod **every 2 minutes**, I could apply a new `Schedule` like:

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: ChaosSchedule
metadata:
  name: schedule-podkill-every-two-minutes
spec:
  schedule:
    cron: "@every 2m"
  chaos:
    podChaos:
      action: container-kill
      mode: one
      selector:
        labelSelectors:
          "app.kubernetes.io/component": "monitor"
```

I want to kill an AWS instance **every 10 minutes**, I could apply a new `Schedule` like:

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: ChaosSchedule
metadata:
  name: schedule-awschaos-ec2-stop-every-five-minutes
spec:
  schedule:
    cron: "@every 5m"
  chaos:
    awsChaos:
      action: ec2-stop
      secretName: "cloud-key-secret"
      awsRegion: "us-east-1"
      ec2Instance: "instance-id-placeholder"
      endpoint: "endpoint-placeholder"
```

> And you might mention that the spec of different Chaos is in different fields: `spec.podChaos` and `spec.awsChaos`, because I prefer to use combination instead of schemaless.

With `PodChaos` and `AwsChaos`, it looks good!

For another type of chaos, I want introduce another concept called `ChaosPeriod` which hold the `duration`.

As we trying to use `reconciler` instead of `webhook`, `reconciler` will do something that could affect the real world that let `status` turns into `spec`. I also want to make some changes to the "common controller". I want to add a new field in some chaos CRD, called `Phase` (or other names), which means the expected state of chaos for now, available values are `Effecting` and `Done`.

For example, I want to inject network delay with `duration:2m` in each 10 minutes:

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: ChaosSchedule
metadata:
  name: schedule-awschaos-ec2-stop-every-five-minutes
spec:
  schedule:
    cron: "@every 10m"
    chaosPeriod:
      duration: 2m
      networkChaos:
        action: delay
        mode: one
        selector:
          labelSelectors:
            "app.kubernetes.io/component": "tikv"
        delay:
          latency: "90ms"
          correlation: "25"
          jitter: "90ms"
        duration: "10s"
```

**Reply by Colstuwjx** [reply]
@STRRL Yes! After removing the `interval` in each Chaos, it looks good!

> There are two types of chaos, one is like NetworkChaos, which could affect for a duration. Another is like AwsChaos, which only executes one operation.

So each **state** chaos like NetworkChaos could use `duration`, and the others like `AwsChaos` **CAN NOT**, right?

> 3-layer-structure, Schedule -> ChaosPeriod -> Chaos

IMO, I think we should make each chaos could works independently, as it's the simplest way to try out chaos mesh. If `chaosPeriod` is required while running a `NetworkChaos` such **state** chaos, it seems a little complicated to the newbies.

In a word, the design of these three components, `ChaosSchedule `, `ChaosPeriod`, `Chaos`(aka MetaChaos, it adds some fields toward the current chaos) is the separation of concerns, I think it should close to the right answer!

**Reply by STRRL** [reply]
> So each state chaos like NetworkChaos could use duration, and the others like AwsChaos CAN NOT, right?

Yes!

> If chaosPeriod is required while running a NetworkChaos such state chaos, it seems a little complicated to the newbies. 

`ChaosPeriod` is not required for running a `NetworkChaos`. Use could apply one `NetworkChaos` for ad-hoc usage.

If a user only applies a `NetworkChaos`, Chaos Mesh will affect the network chaos right now, and recover until set Phase to `Done` or remove it.

If a user applies a `NetworkChaos` wrapped by `ChaosPeriod`, Chaos Mesh will affect chaos right now, and recover after `duration`. It's also suitable for ad-hoc usage.

If a user applies a `NetworkChaos` wrapped by `Schedule`, its behavior is mostly like the current two-phase controller.

**Reply by Colstuwjx** [reply]
@STRRL Awesome! LGTM!

**Reply by YangKeao** [reply]
@STRRL  This configuration is intuitive for the Chaos Mesh developers, but it's hard to understand and use for users. 

> If a user applies a `NetworkChaos` wrapped by `Schedule`, its behavior is mostly like the current two-phase controller.

? Do you mean a `NetworkChaos` wrapped by `ChaosPeriod` wrapped by `Schedule`?

**Reply by STRRL** [reply]
> ? Do you mean a NetworkChaos wrapped by ChaosPeriod wrapped by Schedule?

Yes

**Reply by STRRL** [reply]
> This configuration is intuitive for the Chaos Mesh developers, but it's hard to understand and use for users.

Maybe we could keep the current Kubernetes CRDs. Then migrate it to `Schedule/ChaosPeriod/MetaChaos`, and apply the common controller on MetaChaos.

**Comment by YangKeao** [comment]
@STRRL 

I have another proposal: Keep the `duration` field in the `XXXChaos`, and setup another controller to track this field. I have came up an idea (actually, learn from kubernetes) to solve both this problem and some problems of selector.
As we have agreed that `XXXChaos` should be able to live even after `Recover`, we should unbind the `Recover` and `delete`:

## Unbind `Recover` and `Delete`

The `Chaos` should have two desired phase: one is `Running` and the other is `Finished`, they should be recorded in the `Status`. The `Running` doesn't mean all chaos injection applied. Instead, it could be partially applied (and the rest of them failed). The status of injected `Pod` or `Container` should be recorded in the `Status` to manage the error report and retry (but it's another issue to detail the implementation). After created, the desired phase is set to `Running` by default.

Use complicated `Status` (which will record every injected Pod or Container) but not a single field to represent the "current phase" would also be helpful to extend the `selector` later.

## `duration` controller and `delete` controller

We can setup another two controller to handle the `duration` field and the `delete` behavior from user. The implementation would be really simple: 

The `duration` controller will set desired phase to `Finished` when it arrives `createTime + duration`.
The `delete` controller will set the desired status into `Finished` and check the `Status` to see whether all `Pod`/ containers have been recovered. If not, do nothing, if they were, remove the `finalizer` and let it be deleted.

## `Schedule` CRD

Then the `Schedule` could be a really simple CRD like `CronJob`. We can also add a `historyLimit` field to keep some `XXXChaos` even after the desired of status of them are finished and the `Status` of them seem to be clean enough.

## `pause`

A `Schedule` and the `XXXChaos` could all be paused. After `paused`, the `Schedule` will not start new `XXXChaos` and all started `XXXChaos` will also be paused. When a `XXXChaos` is paused, the desired status of it will be set to `Finished`. We can reference the implementation of `spec.suspend` field in `CronJob`

**Reply by YangKeao** [reply]
The idea of seperated controller for `duration` and `delete` came from the [TTL Controller](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/) of kubernetes, which will delete a finished job after a period of time.

The idea of tracking every pod/container injection status in the `Status` came from `ReplicaSet`. The only difference is that the `ReplicaSet` will create `Pod` as an entity, but `XXXChaos` doesn't have any entity. The "every" should be defined carefully especially considering the "target" selector in `NetworkChaos`. We could discuss about it in another issue.

**Reply by YangKeao** [reply]
Here is a road map to implement this feature:

1. (1-2 week) Modify current two phase controller to fit the "desired phase" and status design. Actually two phase can also be abstracted into a single controller with modifying "desired phase". It could also simplify nearly every chaos implementation by abstracting selector (Fix #961 in a better way, for not only `Recover`, but also `Apply`)
2. (1 week) Add a duration only controller with a standalone controller mentioned in the design.
3. (1 week) Add a `Schedule` CRD and deprecate `scheduler` field.

Only the third step will break compatibility. I believe it will take more time to review than coding :yum: , so I would continue to develop in different branch and submit multiple PRs to avoid blocking my development schedule.

**Reply by YangKeao** [reply]
Any modification and more problem will be filed in this discussion.

**Reply by AsterNighT** [reply]
That is intuitive to me. In fact I'm about to come to a similar idea when I ran into this reply 🍻.
I've got another question.

> A Schedule and the XXXChaos could all be paused. After paused, the Schedule will not start new XXXChaos and all started XXXChaos will also be paused. When a XXXChaos is paused, the desired status of it will be set to Finished. We can reference the implementation of spec.suspend field in CronJob

I can see it is necessary to `pause` a `Schedule`, but do we really need `pause` for a regular `XXXChaos` created by some `Schedule` ? I mean, what if we simply delete the job and create a new one when we resume it? Would that simplify the designing further?
